src/main/java/com/google/step/datamanager/DealVoteCountManagerDatastore.java

```diff
@@ -34,22 +36,30 @@ public int getVotes(long dealId) {
   }
 
   @Override
-  public List<Long> sortDealsInOrderOfVotes(List<Long> dealIds) {
+  public List<Long> sortDealsInOrderOfVotes(List<Long> dealIds, int limit) {
     List<Long> dealIdResults = new ArrayList<>();
     List<Long> dealIdsArrayList = new ArrayList<>(dealIds);
     if (dealIds.size() > 0) {
       Filter dealFilter = new FilterPredicate("deal", FilterOperator.IN, dealIds);
       Query query =
           new Query("DealVote").setFilter(dealFilter).addSort("votes", SortDirection.DESCENDING);
       PreparedQuery pq = datastore.prepare(query);
-      Iterable<Entity> entities = pq.asIterable();
+      Iterable<Entity> entities = null;
+      if (limit == -1) { // Fetch all
+        entities = pq.asIterable();
+      } else {
+        entities = pq.asIterable(FetchOptions.Builder.withLimit(limit));
+      }
       for (Entity entity : entities) {
         dealIdsArrayList.remove(new Long((long) entity.getProperty("deal")));
         dealIdResults.add((long) entity.getProperty("deal"));
       }
+      dealIdResults.addAll(dealIdsArrayList);
     }
-    // Add those that have not been voted on and not in datastore to the end
-    dealIdResults.addAll(dealIdsArrayList);
+    dealIdResults =
+        limit == -1
+            ? dealIdResults
+            : dealIdResults.stream().limit(limit).collect(Collectors.toList());
```

Comment:
Why is this necessary if the fetch options in the prepared query already specified the fetch limit?

Reply:
cos the entities retrieved are only those deals with a vote count, so there are some deals that might not have been voted yet and wont be in the datastore, which is equivalent to a votecount of 0 and will have to be added into the list later on. i changed the fetching to fetch all at first